test/rules/no-null-undefined-union/test.ts.lint

```diff
@@ -0,0 +1,55 @@
+[typescript]: >= 2.4.0
+
+interface someInterface {
+    a: number | undefined | null;
+    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ [0]
+    b: boolean;
+}
+
+const c: string | null | undefined;
+      ~~~~~~~~~~~~~~~~~~~~~~~~~~~~ [0]
+
+export type SomeType =
+~~~~~~~~~~~~~~~~~~~~~~
+    | null
+~~~~~~~~~~
+    | undefined
+~~~~~~~~~~~~~~~
+    | boolean;
+~~~~~~~~~~~~~~ [0]
+
+const someFunc = (): string | undefined | null => {}
+                 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ [0]
+
+const someFunc = (foo: null | string | undefined, bar: boolean) => {}
+                  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ [0]
+
+function someFunc(): number | undefined | null {}
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ [0]
+
+function someFunc(): Promise<number | null | undefined> {} // error
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ [0]
+
+function someFunc(bar: boolean, foo: null | number | undefined) {}
+                                ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ [0]
+
+function someFunc() {
+~~~~~~~~~~~~~~~~~~~~~
+    const somePredicate = (): boolean => true;
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+    const someFunc = (): string | null => null;
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+    let foo;
+~~~~~~~~~~~~
+    if (somePredicate()) {
+~~~~~~~~~~~~~~~~~~~~~~~~~~
+        foo = someFunc();
+~~~~~~~~~~~~~~~~~~~~~~~~~
+    }
+~~~~~
+    return foo;
+~~~~~~~~~~~~~~~
+}
+~ [0]
```

Comment:
this is a really large error range... I don't think you should report the failure on the whole function body. Instead, report it on the function name or near the return value typedef (or where the typedef would go, if it's inferred). consider how tsc reports errors on return values:

![image](https://user-images.githubusercontent.com/723999/55035142-83e24b80-4fed-11e9-8f30-eda054248948.png)

Reply:
```
    if (isSignatureDeclaration(node)) {
        const signature = tc.getSignatureFromDeclaration(node);
        return signature === undefined ? undefined : signature.getReturnType();
    }
```

In this bit, I'm not sure what I can extract off of either `node` or `signature` to get a smaller range.

Reply:
alright, this is fine for now, maybe we can improve this in a future PR